examples/mcmc.jl

```diff
@@ -1,52 +1,100 @@
+
+# Truncated Distributions
+struct TruncNorm <: Gen.Distribution{Float64} end
+const trunc_norm = TruncNorm()
+function Gen.random(::TruncNorm, mu::U, noise::T, low::T, high::T) where {U<:Real,T<:Real}
+    d = Distributions.Truncated(Distributions.Normal(mu, noise),
+                                low, high)
+    return Distributions.rand(d)
+end;
+function Gen.logpdf(::TruncNorm, x::Float64, mu::U, noise::T, low::T, high::T) where {U<:Real,T<:Real}
+    d = Distributions.Truncated(Distributions.Normal(mu, noise),
+                                low, high)
+    return Distributions.logpdf(d, x)
+end;
+
+# this function does not work because the max and min function do not offer an interface for the @trace macro, they would have to be implemented as probability distributions
+@gen function bounded_drift(start::Float64, sd::Float64,
+    lower::Float64=-Inf, upper::Float64=Inf)
+    val ~ normal(start, sd)
+    val_lower_bounded = @trace(max(val, lower), :vl)
+    val_bounded = @trace(min(val_lower_bounded, upper), :vb)
+    return val_bounded
+end
 
-include("__@DIR__/helpers.jl")
-
+# this proposal function implements a truncated random walk for mass and restitution
 @gen function proposal(tr::Gen.Trace)
-    # TODO: fill in TODO's
     # HINT: https://www.gen.dev/tutorials/iterative-inference/tutorial#mcmc-2
     #
     # get previous values from `tr`
-    prev_mass = TODO
-    prev_res = TODO
+    choices = get_choices(tr)
+    prev_mass = choices[:prior => 1 => :mass]
+    prev_res  = choices[:prior => 1 => :restitution]
+    
     # sample new values conditioned on the old ones
-    mass ~ TODO
-    restitution ~ TODO
+    mass = {:prior => 1 => :mass} ~ trunc_norm(prev_mass, .1, 0., Inf)
+    restitution = {:prior => 1 => :restitution} ~ trunc_norm(prev_res, .1, 0., 1.)
+    
     # the return of this function is not
     # neccessary but could be useful
     # for debugging.
     return (mass, restitution)
 end
 
+function get_zs(tr::Gen.Trace)
+    # get the z positions
+    states = get_retval(tr)
+    map(st -> st.kinematics[1].position[3], states)
+end
+
 """
     inference_procedure
 
 Performs Metropolis-Hastings MCMC.
 """
 function inference_procedure(gm_args::Tuple,
-                             obs::Gen.Choicemap,
-                             steps::Int = 100)
+                             obs::Gen.ChoiceMap, 
+                             update_vis,
+                             steps::Int = 20)
 
     # start with an initial guess of physical latents
     # `ls` is the log score or how well this
     # initial guess explains the observations
     tr, ls = Gen.generate(model, gm_args, obs)
+    # visualize predictions of initial trace
+    update_vis(tr)
 
     println("Initial logscore: $(ls)")
 
-    # TODO: use this to count the number of accepted moves
+    # count the number of accepted moves and track accepted proposals
     acceptance_count = 0
+    mass_log = Vector{Float64}()
+    res_log = Vector{Float64}()
```

Comment:
Rather than having to re-size these vectors dynamically, you can initialize them with a fixed size

```julia
v = Vector{Float64}(undef, steps) 

v[i] = value
```

see docs for [`undef`](https://docs.julialang.org/en/v1/base/arrays/#Core.undef)